plugins/chat/app/services/chat/auto_join_channel_batch.rb

```diff
@@ -0,0 +1,112 @@
+# NOTE: When changing auto-join logic, make sure to update the `settings.auto_join_users_info` translation as well.
+# frozen_string_literal: true
+
+module Chat
+  # Service responsible to create memberships for a channel and a section of user ids
+  #
+  # @example
+  #  Chat::AutoJoinChannelBatch.call(
+  #   chat_channel_id: 1,
+  #   starts_at: 27,
+  #   ends_at: 58,
+  #  )
+  #
+  class AutoJoinChannelBatch
+    include Service::Base
+
+    contract
+    model :channel
+    model :category
+    step :create_memberships
+    step :recalculate_user_count
+    step :publish_new_channel
+
+    class Contract
+      attribute :chat_channel_id, :integer
+      validates :chat_channel_id, presence: true
+      attribute :starts_at, :integer
+      attribute :ends_at, :integer
+
+      validate :boundaries
+      def boundaries
+        if %w[starts_at ends_at].all? { |attr| self.public_send(attr).blank? }
+          errors.add :base, "starts_at or ends_at missing"
+        end
+
+        if ends_at && starts_at
+          errors.add(:base, "ends_at can't be higher than starts_at") if ends_at < starts_at
+        end
+      end
+    end
+
+    private
+
+    def fetch_channel(contract:, **)
+      ::Chat::CategoryChannel.find_by(id: contract.chat_channel_id, auto_join_users: true)
+    end
+
+    def fetch_category(channel:, **)
+      channel.chatable
+    end
+
+    def create_memberships(channel:, contract:, category:, **)
+      query_args = {
+        chat_channel_id: channel.id,
+        start: contract.starts_at,
```

Comment:
I know these start/end params are from the old service but probably makes sense to rename them as `start_user_id` and `end_user_id`, otherwise on first glance it seems like they should be start/end dates.

Reply:
Yes this is also something I did at first :p